pkg/operator/etcdcertsigner/etcdcertsignercontroller.go

```diff
@@ -193,108 +218,151 @@ func (c *EtcdCertSignerController) syncAllMasters(recorder events.Recorder) erro
 	return utilerrors.NewAggregate(errs)
 }
 
-// validateCertificateSAN makes sure that a given certificate is valid
-// at least for the SANs defined in the configuration.
-// NOTE: certName argument is used only for forming the error message
-func validateCertificateSAN(certData map[string][]byte, ipAddresses []string, certName string) error {
-	certificates, err := cert.ParseCertsPEM(certData["tls.crt"])
+// ensureCertSecret attempts to ensure the existence of secrets containing the
+// etcd cert (cert+key) pairs needed for an etcd member.
+func (c *EtcdCertSignerController) ensureCertSecrets(node *corev1.Node, recorder events.Recorder) []error {
+	ipAddresses, err := dnshelpers.GetInternalIPAddressesForNodeName(node)
 	if err != nil {
-		return fmt.Errorf("could not parse TLS cert for %s: %w", certName, err)
+		return []error{err}
 	}
-	for _, ipAddress := range ipAddresses {
-		if err := certificates[0].VerifyHostname(ipAddress); err != nil {
-			return fmt.Errorf("SAN for the certificate %s does not include %s: %w", certName, ipAddress, err)
+
+	errs := []error{}
+	for _, certConfig := range certConfigMap {
+		err := c.ensureCertSecret(node, ipAddresses, certConfig, recorder)
+		if err != nil {
+			errs = append(errs, err)
+			continue
 		}
 	}
-	return nil
+	return errs
 }
 
-// Ensure that all secrets exist for the node and validate that the IP addresses are correct.
-func (c *EtcdCertSignerController) ensureAndValidateSecrets(node *corev1.Node, nodeInternalIPs []string, recorder events.Recorder) error {
-	etcdPeerClientSecretName := tlshelpers.GetPeerClientSecretNameForNode(node.Name)
-	etcdServingSecretName := tlshelpers.GetServingSecretNameForNode(node.Name)
-	metricsServingSecretName := tlshelpers.GetServingMetricsSecretNameForNode(node.Name)
+// ensureCertSecret attempts to ensure the existence of a secret containing an
+// etcd cert (cert+key) pair. The secret will be created if it does not
+// exist. If the secret exists but contains an invalid cert pair, it will be
+// updated with a new cert pair.
+func (c *EtcdCertSignerController) ensureCertSecret(node *corev1.Node, ipAddresses []string, certConfig etcdCertConfig, recorder events.Recorder) error {
+	secretName := certConfig.secretNameFunc(node.Name)
 
-	secretNames := []string{etcdPeerClientSecretName, etcdServingSecretName, metricsServingSecretName}
+	secret, err := c.secretLister.Secrets(operatorclient.TargetNamespace).Get(secretName)
+	if err != nil && !errors.IsNotFound(err) {
+		return err
+	}
 
-	errs := []error{}
-	for _, secretName := range secretNames {
-		secret, err := c.secretLister.Secrets(operatorclient.TargetNamespace).Get(secretName)
-		if errors.IsNotFound(err) {
+	nodeUIDUnchanged := false
+	if secret != nil {
+		nodeUIDUnchanged = string(node.UID) == secret.Annotations[nodeUIDAnnotation]
```

Comment:
we have 2 possible conditions here do we need them handled separately wrt to error message?
- no annotation found (possible upgrade and expected)
- uid does not match

Reply:
I will say that if we can prove its safe (should be) rotation of these certs in 4.8 would not be a bad thing 0:)

Reply:
Yes, this is definitely in the direction of rotation support.